src/protocols/rdp/channels/rail.c

```diff
@@ -0,0 +1,197 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "channels/rail.h"
+#include "plugins/channels.h"
+#include "rdp.h"
+#include "settings.h"
+
+#include <freerdp/client/rail.h>
+#include <freerdp/event.h>
+#include <freerdp/freerdp.h>
+#include <freerdp/rail.h>
+#include <guacamole/client.h>
+#include <winpr/wtypes.h>
+#include <winpr/wtsapi.h>
+
+#include <stddef.h>
+#include <string.h>
+
+/**
+ * Completes initialization of the RemoteApp session, sending client system
+ * parameters and executing the desired RemoteApp command using the Client
+ * System Parameters Update PDU and Client Execute PDU respectively. These PDUs
+ * MUST be sent for the desired RemoteApp to run, and MUST NOT be sent until
+ * after a Handshake or HandshakeEx PDU has been received. See:
+ *
+ * https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdperp/60344497-883f-4711-8b9a-828d1c580195 (System Parameters Update PDU)
+ * https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdperp/98a6e3c3-c2a9-42cc-ad91-0d9a6c211138 (Client Execute PDU)
+ *
+ * @param rail
+ *     The RailClientContext structure used by FreeRDP to handle the RAIL
+ *     channel for the current RDP session.
+ *
+ * @return
+ *     CHANNEL_RC_OK (zero) if the PDUs were sent successfully, an error code
+ *     (non-zero) otherwise.
+ */
+static UINT guac_rdp_rail_complete_handshake(RailClientContext* rail) {
+
+    guac_client* client = (guac_client*) rail->custom;
+    guac_rdp_client* rdp_client = (guac_rdp_client*) client->data;
+
+    RAIL_SYSPARAM_ORDER sysparam = {
+        .workArea = {
+            .left   = 0,
+            .top    = 0,
+            .right  = rdp_client->settings->width,
+            .bottom = rdp_client->settings->height
+        },
+        .dragFullWindows = FALSE
+    };
+
+    /* Send client system parameters */
+    UINT status = rail->ClientSystemParam(rail, &sysparam);
+    if (status != CHANNEL_RC_OK)
+        return status;
+
+    RAIL_EXEC_ORDER exec = {
+        .RemoteApplicationProgram = rdp_client->settings->remote_app,
+        .RemoteApplicationWorkingDir = rdp_client->settings->remote_app_dir,
+        .RemoteApplicationArguments = rdp_client->settings->remote_app_args,
+    };
+
+    /* Execute desired RemoteApp command */
+    return rail->ClientExecute(rail, &exec);
+
+}
+
+/**
+ * Callback which is invoked when a Handshake PDU is received from the RDP
+ * server. No communication for RemoteApp may occur until the Handshake PDU
+ * (or, alternatively, the HandshakeEx PDU) is received. See:
+ *
+ * https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdperp/cec4eb83-b304-43c9-8378-b5b8f5e7082a
+ *
+ * @param rail
+ *     The RailClientContext structure used by FreeRDP to handle the RAIL
+ *     channel for the current RDP session.
+ *
+ * @param handshake
```

Comment:
This isn't used for anything?

Reply:
Not currently, no. We ignore the contents of the handshake, simply using it to signal when client system parameters should be sent and execution of the RemoteApp should be requested.

This is actually in line with the behavior of xfreerdp, which also ignores the contents of the handshake PDU received from the server:

https://github.com/FreeRDP/FreeRDP/blob/02614cce49ce6ff120a1306962c5d3ff6b05ea75/client/X11/xf_rail.c#L892-L901

Reply:
Huh, ok.